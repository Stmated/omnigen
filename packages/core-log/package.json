{
  "name": "@omnigen/core-log",
  "version": "0.0.0",
  "private": true,
  "license": "MIT",
  "type": "module",
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "scripts": {
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "dev": "tsc -w",
    "lint": "TIMING=1 eslint src/**/*.ts* --fix",
    "test": "vitest run",
    "test:watch": "vitest"
  },
  "dependencies": {
    "pino": "8.6.0"
  },
  "devDependencies": {
    "@omnigen/config-typescript": "workspace:*",
    "@omnigen/eslint-config-custom": "workspace:*",
    "colors": "^1.4.0",
    "typescript": "^5.5.0-beta",
    "pino-pretty": "^10.3.1"
  }
}
