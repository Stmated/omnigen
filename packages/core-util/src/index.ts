
export * from './equality/index.js';
export * from './interpret/index.js';
export * from './options/index.js';
export * from './parse/index.js';
export * from './util/index.js';
export * from './visit/index.js';
export * from './write/index.js';
