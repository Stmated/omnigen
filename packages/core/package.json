{
  "name": "@omnigen/core",
  "version": "0.0.0",
  "private": true,
  "license": "MIT",
  "type": "module",
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "scripts": {
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "dev": "tsc -w",
    "lint": "TIMING=1 eslint src/**/*.ts* --fix",
    "test": "vitest run",
    "test:watch": "vitest"
  },
  "dependencies": {
    "@omnigen/core-log": "workspace:*",
    "@etclabscore/string-json-template-language": "^1.0.0",
    "@open-rpc/schema-utils-js": "^1.16.1",
    "change-case": "^4.1.2",
    "deepmerge": "^4.3.0",
    "js-yaml": "^4.1.0",
    "json-pointer": "^0.6.2",
    "node-fetch": "^2.6.9",
    "object-hash": "^3.0.0",
    "string-similarity": "^4.0.4",
    "ts-json-schema-generator": "^1.2.0",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@omnigen/config-typescript": "workspace:*",
    "@omnigen/eslint-config-custom": "workspace:*",
    "@swc/core": "^1.3.32",
    "@types/json-pointer": "^1.0.31",
    "@types/json-schema": "^7.0.15",
    "@types/node": "^18.13.0",
    "@types/node-fetch": "^2.6.2",
    "@types/object-hash": "^2.2.1",
    "@types/string-similarity": "^4.0.0",
    "colors": "^1.4.0",
    "nock": "^13.3.0",
    "typescript": "^5.4.2",
    "type-fest": "^3.6.1"
  }
}
