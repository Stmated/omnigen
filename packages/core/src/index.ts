import {PluginAutoRegistry} from '@omnigen/core-plugin';
import {CommonTransformPlugin, CommonTransform2Plugin, fileWriter, CorePlugin} from './CoreUtilPluginInit.ts';

export * from './ast';
export * from './equality';
export * from './parse';
export * from './util';
export * from './visit';
export * from './write';
export * from './CoreUtilPluginInit.ts';

export default PluginAutoRegistry.register([CorePlugin, CommonTransformPlugin, CommonTransform2Plugin, fileWriter]);
