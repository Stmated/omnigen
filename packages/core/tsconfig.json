{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@omnigen/config-typescript/build.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "baseUrl": "./src"
  }
}
