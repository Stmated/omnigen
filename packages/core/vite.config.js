import {resolve} from 'path';
import {defineConfig} from 'vite';

export default defineConfig({
  build: {
    outDir: 'dist/lib',
    lib: {
      // Could also be a dictionary or array of multiple entry points
      entry: resolve(__dirname, 'src/index.ts'),
      name: '@omnigen/core',
      // the proper extensions will be added
      // fileName: 'core',
    },
    rollupOptions: {
      // make sure to externalize deps that shouldn't be bundled into your library
      external: ['node-fetch', 'promisify', 'node:fs', 'node:url', 'find-up', 'fs/promises', 'fs'],
      output: {
        // Provide global variables to use in the UMD build for externalized deps
        // globals: {
        //   vue: 'Vue',
        // },
      },
      treeshake: 'safest',
    },
  },
  optimizeDeps: {
    exclude: ['sync-fetch'],
  },
});
