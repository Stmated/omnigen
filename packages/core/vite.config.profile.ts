import {resolve} from 'path';
import {type UserConfig} from 'vite';

export default {
  build: {
    outDir: 'dist/profile',
    minify: false,
    terserOptions: {
      compress: false,
      mangle: false,
    },
    lib: {
      // Could also be a dictionary or array of multiple entry points
      entry: resolve(__dirname, 'src/index.profiling.ts'),
      name: '@omnigen/core',
      // the proper extensions will be added
      // fileName: 'core',
    },
    rollupOptions: {
      external: ['node-fetch', 'promisify', 'node:fs', 'node:url', 'find-up'],
      output: {
        // Provide global variables to use in the UMD build for externalized deps
        // globals: {
        //   vue: 'Vue',
        // },
      },
    },
  },
} satisfies UserConfig;
