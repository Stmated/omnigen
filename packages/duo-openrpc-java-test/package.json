{
  "name": "@omnigen/duo-openrpc-java-test",
  "version": "0.0.0",
  "private": true,
  "license": "MIT",
  "type": "module",
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "scripts": {
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "dev": "tsc -w",
    "lint": "TIMING=1 eslint src/**/*.ts* --fix",
    "test": "NODE_OPTIONS=--experimental-vm-modules jest --passWithNoTests",
    "test:watch": "jest --watch --passWithNoTests"
  },
  "jest": {
    "preset": "@omnigen/config-jest"
  },
  "dependencies": {
    "@omnigen/core": "0.0.0",
    "@omnigen/core-log": "0.0.0",
    "@omnigen/core-util": "0.0.0",
    "@omnigen/parser-jsonschema": "0.0.0",
    "@omnigen/parser-openrpc": "0.0.0",
    "@omnigen/target-java": "0.0.0",
    "json-schema": "^0.4.0"
  },
  "devDependencies": {
    "@types/jest": "^29.4.0",
    "@types/json-schema": "^7.0.11",
    "@omnigen/config-jest": "0.0.0",
    "@omnigen/config-typescript": "0.0.0",
    "@omnigen/eslint-config-custom": "0.0.0",
    "@omnigen/utils-test": "0.0.0",
    "@omnigen/utils-test-target-java": "0.0.0",
    "typescript": "beta",
    "jest": "^29.4.1",
    "java-parser": "^2.0.3"
  }
}
