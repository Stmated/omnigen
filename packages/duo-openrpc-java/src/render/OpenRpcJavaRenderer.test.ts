import * as fs from 'fs';
import * as path from 'path';
import * as JavaParser from 'java-parser';
import {DEFAULT_TEST_JAVA_OPTIONS, DEFAULT_TEST_TARGET_OPTIONS, JavaTestUtils, OpenRpcTestUtils} from '@omnigen/duo-openrpc-java-test';
import {ParsedJavaTestVisitor} from '@omnigen/utils-test-target-java';
import {describe, expect, test, vi} from 'vitest';
import {DEFAULT_MODEL_TRANSFORM_OPTIONS, DEFAULT_PARSER_OPTIONS, ModelTransformOptions, OmniPrimitiveKind} from '@omnigen/core';
import {ZodCompilationUnitsContext} from '@omnigen/core-util';
import {LoggerFactory} from '@omnigen/core-log';
import {SerializationConstructorAnnotationMode, SerializationLibrary, SerializationPropertyNameMode} from '@omnigen/target-java';

const logger = LoggerFactory.create(import.meta.url);

describe('Java Rendering', () => {

  test('renderAll', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    // TODO: Do this both with and without compressing types -- create different versions and output them in a structure!
    // TODO: Would it be crazy or cool to take ALL the types of all the models and create one HUGE output with common types?

    for (const schemaName of OpenRpcTestUtils.getKnownSchemaNames()) {

      let fileNames: string[];
      try {
        fileNames = await OpenRpcTestUtils.listExampleFileNames(schemaName);
      } catch (ex) {
        throw new Error(`Could not list filenames inside ${schemaName}: ${ex}`, {cause: ex instanceof Error ? ex : undefined});
      }

      for (const fileName of fileNames) {

        try {

          logger.info(`Will render ${fileName}`);
          const result = await JavaTestUtils.getResultFromFilePath(
            `../parser-openrpc/examples/${fileName}`, {}, ZodCompilationUnitsContext,
          );

          let baseDir: string;

          try {
            baseDir = path.resolve(`./.target_test/${schemaName}/${path.basename(fileName, path.extname(fileName))}`);
          } catch (ex) {
            throw new Error(`Could not resolve path of ${fileName}`, {cause: ex});
          }

          for (const cu of result.compilationUnits) {

            if (cu.fileName.indexOf('#') !== -1) {
              throw new Error(`'#' not allowed in CU '${cu.fileName}'`);
            }

            const outDir = `${baseDir}/${cu.directories.join('/')}`;
            if (!fs.existsSync(outDir)) {
              fs.mkdirSync(outDir, {recursive: true});
            }

            const outPath = `${outDir}/${cu.fileName}`;

            try {
              fs.writeFileSync(outPath, cu.content);
            } catch (ex) {
              throw new Error(`Could not write '${outPath}' of '${fileName}': ${ex}`, {cause: ex});
            }

            let cst: JavaParser.CstNode;
            try {
              cst = JavaParser.parse(cu.content);
              expect(cst).toBeDefined();
            } catch (ex) {
              throw new Error(`Could not parse '${schemaName}' '${fileName}' in '${outPath}': ${ex}`, {cause: ex});
            }

            // Visit the syntax tree, but do nothing with the result.
            // It is only to see if any crashes can be found or not.
            const visitor = new ParsedJavaTestVisitor();
            visitor.visit(cst);
          }
        } catch (ex) {
          throw LoggerFactory.formatError(ex, `File ${fileName}`);
        }
      }
    }
  }, {
    timeout: 20_000,
  });

  test('Test multiple inheritance (interfaces)', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const fileContents = await JavaTestUtils.getFileContentsFromFile('multiple-inheritance.json', {
      javaOptions: {...DEFAULT_TEST_JAVA_OPTIONS, serializationLibrary: SerializationLibrary.JACKSON},
    });

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, fileContent] of fileContents) {
      expect(fileContent).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('Type compressions', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    // Check that the property 'common' from A and B are moved into Abs.

    const fileContents = await JavaTestUtils.getFileContentsFromFile('compressable-types.json', {
      modelTransformOptions: {...DEFAULT_MODEL_TRANSFORM_OPTIONS, generifyTypes: false},
      javaOptions: {
        ...DEFAULT_TEST_JAVA_OPTIONS,
        preferNumberType: OmniPrimitiveKind.DOUBLE,
        serializationLibrary: SerializationLibrary.JACKSON,
        serializationPropertyNameMode: SerializationPropertyNameMode.IF_REQUIRED,
      },
    });

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, fileContent] of fileContents) {
      expect(fileContent).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('Enum', async ({task}) => {

    const fileContents = await JavaTestUtils.getFileContentsFromFile('enum.json', {
      javaOptions: {
        ...DEFAULT_TEST_JAVA_OPTIONS,
        includeGeneratedAnnotation: false,
        serializationLibrary: SerializationLibrary.JACKSON,
        serializationPropertyNameMode: SerializationPropertyNameMode.IF_REQUIRED,
      }
    });
    // TODO: Get this to work again -- something is very off with the way things are created -- maybe a miss-mash with the ids and simplification of oneOf -> allOf -> inlined
    //       Might be that we should not inline early, and let it be up to the target language how a type that is only one allOf should be rendered. If it is anonymous it could be inlined always;

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, fileContent] of fileContents) {
      expect(fileContent).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('AdditionalProperties', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const fileContents = await JavaTestUtils.getFileContentsFromFile('additional-properties.json', {
      targetOptions: {...DEFAULT_TEST_TARGET_OPTIONS, additionalPropertiesInterfaceAfterDuplicateCount: 1},
      javaOptions: {
        ...DEFAULT_TEST_JAVA_OPTIONS,
        serializationLibrary: SerializationLibrary.JACKSON,
        serializationPropertyNameMode: SerializationPropertyNameMode.IF_REQUIRED,
      }
    });

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, fileContent] of fileContents) {
      expect(fileContent).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  test('PetStore-Expanded-ClassNames', async () => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const fileContents = await JavaTestUtils.getFileContentsFromFile('petstore-expanded.json');
    const filenames = [...fileContents.keys()];

    expect(filenames).toContain('Pet.java');
    expect(filenames).toContain('DeletePetByIdResponse.java');
    expect(filenames).not.toContain('Pet1.java');
  });

  test('primitive-generics', async () => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const fileContents = await JavaTestUtils.getFileContentsFromFile('primitive-generics.json', {
      javaOptions: {...DEFAULT_TEST_JAVA_OPTIONS, preferNumberType: OmniPrimitiveKind.DOUBLE, serializationLibrary: 'POJO'},
    });

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    expect(fileContents.get('GiveNumberGetCharRequestParams.java')).toMatchSnapshot();

    // TODO: Add more exact checks for all generic types, making sure they have the correct amount of generics and whatever
  });

  test('generic_params', async ({task}) => {

    vi.useFakeTimers({now: new Date('2000-01-02T03:04:05.000Z')});

    const result = await JavaTestUtils.getResultFromFilePath(`./examples/generic_params.json`, {
      parserOptions: {...DEFAULT_PARSER_OPTIONS, defaultAdditionalProperties: false},
      targetOptions: {
        ...DEFAULT_TEST_TARGET_OPTIONS,
        compressSoloReferencedTypes: true,
        compressUnreferencedSubTypes: true,
        additionalPropertiesInterfaceAfterDuplicateCount: 100,
      },
      javaOptions: {
        ...DEFAULT_TEST_JAVA_OPTIONS,
        serializationLibrary: SerializationLibrary.JACKSON,
        serializationPropertyNameMode: SerializationPropertyNameMode.ALWAYS,
      },
    }, ZodCompilationUnitsContext);

    const fileContents = await JavaTestUtils.cuToContentMap(result.compilationUnits);

    expect([...fileContents.keys()].sort()).toMatchSnapshot();
    for (const [fileName, fileContent] of fileContents) {
      expect(fileContent).toMatchFileSnapshot(`./__snapshots__/${task.suite.name}/${task.name}/${fileName}`);
    }
  });

  // TODO: Create test case that actually needs the expandedGenericSourceIdentifier inside GenericsModelTransformer -- it is a good feature, just needs some fixes
});
