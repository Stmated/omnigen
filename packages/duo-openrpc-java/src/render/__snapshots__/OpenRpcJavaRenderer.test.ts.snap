// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Java Rendering > AdditionalProperties 1`] = `
[
  "ErrorUnknown.java",
  "ErrorUnknownError.java",
  "IAdditionalProperties.java",
  "JsonRpcError.java",
  "JsonRpcErrorResponse.java",
  "JsonRpcRequest.java",
  "JsonRpcRequestParams.java",
  "JsonRpcResponse.java",
  "ListThingsRequest.java",
  "ListThingsRequestParams.java",
  "ListThingsResponse.java",
  "Thing.java",
]
`;

exports[`Java Rendering > Enum 1`] = `
[
  "ErrorUnknown.java",
  "ErrorUnknownError.java",
  "JsonRpcError.java",
  "JsonRpcErrorResponse.java",
  "JsonRpcRequest.java",
  "JsonRpcRequestParams.java",
  "JsonRpcResponse.java",
  "ListThingsError100.java",
  "ListThingsError100Error.java",
  "ListThingsRequest.java",
  "ListThingsRequestParams.java",
  "ListThingsResponse.java",
  "Species.java",
  "Tag.java",
  "TagCopy.java",
  "TagOrSpeciesOrString.java",
  "Thing.java",
  "ThingType.java",
]
`;

exports[`Java Rendering > Test multiple inheritance (interfaces) 1`] = `
[
  "A.java",
  "AXOrB.java",
  "Abs.java",
  "B.java",
  "C.java",
  "ErrorUnknown.java",
  "ErrorUnknownError.java",
  "GiveInGetOut2Request.java",
  "GiveInGetOut2RequestParams.java",
  "GiveInGetOut2Response.java",
  "GiveInGetOutRequest.java",
  "GiveInGetOutRequestParams.java",
  "GiveInGetOutResponse.java",
  "IB.java",
  "IC.java",
  "In.java",
  "JsonRpcError.java",
  "JsonRpcErrorResponse.java",
  "JsonRpcRequest.java",
  "JsonRpcRequestParams.java",
  "JsonRpcResponse.java",
  "Out.java",
  "Out2.java",
]
`;

exports[`Java Rendering > Type compressions 1`] = `
[
  "A.java",
  "Abs.java",
  "B.java",
  "ErrorUnknown.java",
  "ErrorUnknownError.java",
  "GiveIn1GetOut1Request.java",
  "GiveIn1GetOut1RequestParams.java",
  "GiveIn1GetOut1Response.java",
  "GiveIn2GetOut2Request.java",
  "GiveIn2GetOut2RequestParams.java",
  "GiveIn2GetOut2Response.java",
  "In1.java",
  "In2.java",
  "JsonRpcError.java",
  "JsonRpcErrorResponse.java",
  "JsonRpcRequest.java",
  "JsonRpcRequestParams.java",
  "JsonRpcResponse.java",
]
`;

exports[`Java Rendering > generic_params 1`] = `
[
  "AbstractRequestData.java",
  "DepositRequest.java",
  "DepositResponse.java",
  "ErrorUnknown.java",
  "ErrorUnknownError.java",
  "JsonRpcError.java",
  "JsonRpcErrorResponse.java",
  "JsonRpcRequest.java",
  "JsonRpcRequestParams.java",
  "JsonRpcResponse.java",
  "RefundRequest.java",
  "RefundResponse.java",
]
`;

exports[`Java Rendering > primitive-generics 1`] = `
[
  "ErrorUnknown.java",
  "ErrorUnknownError.java",
  "GiveIntGetDoubleRequest.java",
  "GiveIntGetDoubleRequestParams.java",
  "GiveIntGetDoubleResponse.java",
  "GiveNumberGetCharRequest.java",
  "GiveNumberGetCharRequestParams.java",
  "GiveNumberGetCharResponse.java",
  "GiveStringGetStringRequest.java",
  "GiveStringGetStringRequestParams.java",
  "GiveStringGetStringResponse.java",
  "JsonRpcError.java",
  "JsonRpcErrorResponse.java",
  "JsonRpcRequest.java",
  "JsonRpcRequestParams.java",
  "JsonRpcResponse.java",
]
`;

exports[`Java Rendering > primitive-generics 2`] = `
"package generated.omnigen;

import javax.annotation.Generated;

@Generated(value = "omnigen", date = "2000-01-02T03:04:05.000Z")
public class GiveNumberGetCharRequestParams extends JsonRpcRequestParams<Double> {
  public GiveNumberGetCharRequestParams(double param) {
    super(param);
  }
}
"
`;
